docs/investigations/graphql-partial-updates/graphql-partial-updates.md

```diff
@@ -0,0 +1,156 @@
+# GraphQL Partial Updates
+
+## Overview
+
+This documents describes and compares different approaches to designing partial updates in graphql schema. Most of presented approaches have a link to an example they were based on. 
+
+## Partial update mutation designs
+
+### 1. Simple Patch
+
+#### Schema
+```graphql
+input ApplicationPatchInput {
+    name: String
+    description: String
+    webhooks: [WebhookInput!]
+    healthCheckURL: String
+    apis: [APIDefinitionInput!]
+    eventAPIs: [EventAPIDefinitionInput!]
+    documents: [DocumentInput!]
+}
+
+type Mutation {
+    updateApplication(id: ID!, patch: ApplicationPatchInput!): Application!
+}
+```
+
+#### Description
+In this approach every field has to be optional. If the field is defined, we use the passed value to set the new value. If the field is not defined we don't change the current value.
```

Comment:
For string, this approach seems to be ok, but I am wondering about fields that store for example integer, is `0` correct value? Let assume that we have `size` for runtime :) representing the number of nodes.
- 0 can mean it was scaled down to 0 nodes
- 0 can mean that we don't know

and I am not convinced that this is ok.

Reply:
If such a case occurred, we came up with several ideas:
First, we thought about providing a wrapper type which would look like that:
```
input nilableInteger {
	value: Int
	isNil: Boolean
}
```
and the size would be of type `nilableInteger`.
In the case when we don't know the value, we would set `isNil` value to true. Otherwise, we would set `isNil` to false and `value` to 0.

Another idea would be that we stick to strings in our api regardless of what value would suggest. So the `size` value would be a string.

Also, we investigated more the `Commands & Actions` approach, where such a problem wouldn't occur, but we would have to specify setter inputs for pretty much all of inputs.

There is also a case to consider with default values. For example, for now, we have `VersionInput` type which has two boolean values with default values. Considering a case when we want to update just value of the version type, and actual booleans differ from the default values, they would've been overridden. One of our propositions to handle that case is to create another input type for the `Version`.

We have to remember that if we want to use `PATCH` updates, we'll have to specify additional inputs for each existing `PUT` input